### PR TITLE
Enabling SSO via applink in sample app

### PR DESCRIPTION
**Pull request by lalwani**
Description:
Updated the sample app for authentication flow using app links

- Added a new button to login via uber app using applink
- Generating code challenge, code verifier pair and sending code challenge as part of the sso request
- Once the sso flow is successfully complete using applink we get back auth code. Using this auth code and code verifier to get auth tokens

### CODE REVIEW

File: samples/login-sample/src/main/java/com/uber/sdk/android/samples/LoginSampleActivity.java [hunk 1]
```diff
@@ -127,14 +140,35 @@ protected void onCreate(Bundle savedInstanceState) {
...
             @Override
             public void onClick(View v) {
                 loginManager.login(LoginSampleActivity.this);
             }
         });
+        appLinkButton = findViewById(R.id.applink_uber_button);
+        appLinkButton.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View v) {
+                Intent intent = new Intent(Intent.ACTION_VIEW);
```
Comment (tyvsmith) [inline samples/login-sample/src/main/java/com/uber/sdk/android/samples/LoginSampleActivity.java:154]: This seems like more boilerplate than we may want a dev to write. See loginManager.login as an example where we abstracted startActivityForResult handling. It also provides a proxy activity to catch and return callbacks. We probably want to do something similar.

File: samples/login-sample/src/main/java/com/uber/sdk/android/samples/LoginSampleActivity.java [hunk 2]
```diff
@@ -127,14 +140,35 @@ protected void onCreate(Bundle savedInstanceState) {
...
+                }
+                intent.setData(
+                        AuthUriAssembler.assemble(
+                                        ResponseType.CODE,
+                                        loginManager.getSessionConfiguration(),
+                                        codeChallenge,
+                                        null
+                                )
+                );
+                startActivityForResult(intent, CUSTOM_BUTTON_REQUEST_CODE);
```
Comment (tyvsmith) [inline samples/login-sample/src/main/java/com/uber/sdk/android/samples/LoginSampleActivity.java:169]: Do we need new explicit handling for onActivityResult if you're using a new call here?

File: samples/login-sample/src/main/java/com/uber/sdk/android/samples/LoginSampleActivity.java [hunk 3]
```diff
@@ -179,9 +213,23 @@ public void onLoginSuccess(@NonNull AccessToken accessToken) {
...
         public void onAuthorizationCodeReceived(@NonNull String authorizationCode) {
-            Toast.makeText(LoginSampleActivity.this, getString(R.string.authorization_code_message, authorizationCode),
-                    Toast.LENGTH_LONG)
-                    .show();
+            new AuthorizationCodeGrantFlow(loginManager.getSessionConfiguration(),
+                    authorizationCode,
+                    CODE_VERIFIER).execute(new TokenRequestFlow.TokenRequestFlowCallback() {
+                @Override
+                public void onSuccess(AccessToken accessToken) {
+                    new AccessTokenManager(LoginSampleActivity.this).setAccessToken(accessToken);
```
Comment (tyvsmith) [inline samples/login-sample/src/main/java/com/uber/sdk/android/samples/LoginSampleActivity.java:221]: You'd probably want your abstract I mentioned above to handle this, as opposed to the user doing it in callback

File: samples/login-sample/src/main/java/com/uber/sdk/android/samples/LoginSampleActivity.java (hunk 1 above)
Comment (lalwani) [reply samples/login-sample/src/main/java/com/uber/sdk/android/samples/LoginSampleActivity.java:154]: This is intentionally done here.
With applink support devs may not use the SDK to launch single sign on with Uber app. So, this piece is added to guide the devs how they can assemble the applink and use startActivityForResult to launch it

File: samples/login-sample/src/main/java/com/uber/sdk/android/samples/LoginSampleActivity.java (hunk 3 above)
Comment (lalwani) [reply samples/login-sample/src/main/java/com/uber/sdk/android/samples/LoginSampleActivity.java:221]: All the changes have been made assuming the developer is not integrating the SDK rather using the assembling applink directly in their app and launching it.

Comment (tyvsmith) [reply samples/login-sample/src/main/java/com/uber/sdk/android/samples/LoginSampleActivity.java:221]: Got it, then I think this is a bit confusing and doesn't communicate usage well as the example here still uses the SDK to acccess the TokenManager, Assemberl, and other classes, adds some custom logic, and intermixes a different usage.

If you're intention is to have an example that could entirely get away from using the SDK, then it's probably worth doing a second sample app here without the dep on the main SDK and possibly extracting the needed classes into a separate module (separate deploying maven artifact) that they can depend on without the rest.

File: core-android/src/main/java/com/uber/sdk/android/core/auth/AuthUriAssembler.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+import androidx.annotation.NonNull;
+
+import com.uber.sdk.android.core.BuildConfig;
+import com.uber.sdk.android.core.utils.AppProtocol;
+import com.uber.sdk.core.client.SessionConfiguration;
+
+import java.util.Locale;
+
+public class AuthUriAssembler {
```
Comment (tyvsmith) [inline core-android/src/main/java/com/uber/sdk/android/core/auth/AuthUriAssembler.java:14]: FYI, these are in the SDK components, not sample app. Was that intended if this is to show an SDK less version?

Comment (lalwani) [reply core-android/src/main/java/com/uber/sdk/android/core/auth/AuthUriAssembler.java:14]: Yes, it was intended. Basically I want to put out 2 samples
1. applink usage with sdk
2. applink usage without sdk

AuthUriAssembler would stay inside the sdk and will be used with #1. 

For #2 I will create a new sample app, as you recommended, to keep things cleaner.

File: samples/login-sample/src/main/java/com/uber/sdk/android/samples/LoginSampleActivity.java (hunk 3 above)
Comment (lalwani) [reply samples/login-sample/src/main/java/com/uber/sdk/android/samples/LoginSampleActivity.java:221]: fair point. 
i will update the current diff with sdk usage and put up a second PR for another sample app demo-ing usage without android sdk

**Comment by lalwani** [comment]
Closing this one over https://github.com/uber/rides-android-sdk/pull/203

**Pull request closed without merging**